cardano-api/src/Cardano/Api/Tx.hs

```diff
@@ -586,22 +589,29 @@ makeShelleyBootstrapWitness _ ByronTxBody{} _ =
 
 makeShelleyBootstrapWitness nwOrAddr (ShelleyTxBody era txbody _) sk =
     case era of
-      ShelleyBasedEraShelley -> makeShelleyBasedBootstrapWitness era
-                                  nwOrAddr txbody sk
-      ShelleyBasedEraAllegra -> makeShelleyBasedBootstrapWitness era
-                                  nwOrAddr txbody sk
-      ShelleyBasedEraMary    -> makeShelleyBasedBootstrapWitness era
-                                  nwOrAddr txbody sk
-
-makeShelleyBasedBootstrapWitness :: forall era ledgerera.
+      ShelleyBasedEraShelley ->
+        makeShelleyBasedBootstrapWitness
+          (Proxy :: Proxy (Ledger.ShelleyEra StandardCrypto))
+          era nwOrAddr txbody sk
+      ShelleyBasedEraAllegra ->
+        makeShelleyBasedBootstrapWitness
+          (Proxy :: Proxy (Ledger.AllegraEra StandardCrypto))
+          era nwOrAddr txbody sk
+      ShelleyBasedEraMary    ->
+        makeShelleyBasedBootstrapWitness
+          (Proxy :: Proxy (Ledger.MaryEra StandardCrypto))
+          era nwOrAddr txbody sk
+
+makeShelleyBasedBootstrapWitness :: forall era ledgerera proxy.
                                     Shelley.ShelleyBased ledgerera
                                  => Ledger.Crypto ledgerera ~ StandardCrypto
-                                 => ShelleyBasedEra era
+                                 => proxy ledgerera
```

Comment:
So we have this type family:  https://github.com/input-output-hk/cardano-node/blob/master/cardano-api/src/Cardano/Api/Eras.hs#L311
```
type family ShelleyLedgerEra era where

  ShelleyLedgerEra ShelleyEra = Ledger.StandardShelley
  ShelleyLedgerEra AllegraEra = Ledger.StandardAllegra
  ShelleyLedgerEra MaryEra    = Ledger.StandardMary
```

That should allow you to avoid having to use a proxy @redxaxder

Reply:
There's some kind of subtlety involved here that I don't understand. In the original version which didn't have the proxy, ghc was unable to infer the right types. Changing it back brings the error back.

```
src/Cardano/Api/Tx.hs:602:37: error:
    • Could not deduce: Ledger.Crypto ledgerera0 ~ StandardCrypto
      from the context: (Shelley.ShelleyBased ledgerera,
                         Ledger.Crypto ledgerera ~ StandardCrypto)
        bound by the type signature for:
                   makeShelleyBasedBootstrapWitness :: forall era ledgerera.
                                                       (Shelley.ShelleyBased ledgerera,
                                                        Ledger.Crypto ledgerera ~
StandardCrypto) =>
                                                       ShelleyBasedEra era
                                                       -> WitnessNetworkIdOrByronAddres
s
                                                       -> Ledger.TxBody ledgerera
                                                       -> SigningKey ByronKey
                                                       -> Witness era
        at src/Cardano/Api/Tx.hs:(602,37)-(609,47)
      The type variable ‘ledgerera0’ is ambiguous
    • In the ambiguity check for ‘makeShelleyBasedBootstrapWitness’
      To defer the ambiguity check to use sites, enable AllowAmbiguousTypes
      In the type signature:
        makeShelleyBasedBootstrapWitness :: forall era ledgerera.
                                            Shelley.ShelleyBased ledgerera =>
                                            Ledger.Crypto ledgerera ~ StandardCrypto =>
                                            ShelleyBasedEra era
                                            -> WitnessNetworkIdOrByronAddress
                                               -> Ledger.TxBody ledgerera
                                                  -> SigningKey ByronKey -> Witness era
    |
602 | makeShelleyBasedBootstrapWitness :: forall era ledgerera.
    |                                     ^^^^^^^^^^^^^^^^^^^^^...
```

Reply:
The confusing thing to me is that it doesn't look like any of the the ledger changes should have affected type inference here, and this file must have originally compiled without AllowAmbiguousTypes.

Reply:
So the initial change you need to make is: https://github.com/input-output-hk/cardano-node/commit/7575f939ff0bc45372f0b4493ae4e4efd8bb5f09#diff-dd9d25b805092a4cd8fc04864deca453f3c32f3282643af7dc670c2738cd24b1R607